### PR TITLE
Overhaul TUI

### LINKED ISSUE #42: TUI Improvements
- [x] Add Curses file selector in place of curses_get_user_input() for file selections. See #99 and https://github.com/FaithLife-Community/LogosLinuxInstaller/pull/99/commits/fc44b009f91366ec6e67a042c8b04b42bc8b5cd7
- [x] Create a tui_screen.TextScreen (see #99) that shows the list of required packages and show them being scanned/checked off when found, as a form of user feedback, during the dependency checking.
- [x] Add a status line. See #99.
  - When the TUI is running, we cannot see the CLI itself. We may want to consider making the TUI a full app, as in do not return to the CLI but keep the TUI open even on a blank screen, so as not to obscure output.
- [x] Add a progress bar. See #85.
- [x] Rework using curses.wrapper() for error handling. See https://github.com/FaithLife-Community/LogosLinuxInstaller/commit/818358897db2c8f1b95c47212226defc8952adc9.
- [x] Add a progress spinner/wheel to show that we are still working and not stalled. See #99.
- [x] Reduce number of screens by having multiple choices on one??? Probably too painful to generate nicely.
- [x] Implement doupdate instead of refresh

### PR DESCRIPTION
**Pull request by thw26**
This PR overhauls the TUI. It involves some major code restructuring in order to be able to do this. Notably, the code is now much more readable and reusable. I'm sure there are further gains to be had in code reuse.

There may be other needs down the road.

Fixed #42, #102, #104

### CODE REVIEW

**Comment by thw26** [comment]
Some early feedback:

>So we need a check then in the TUI code to determine someone's dialog version to make sure it is at least a particular version, then we need to change the dialog flag to be a config option…

**Comment by thw26** [comment]
Current conflicts exist because I cherry-picked a commit for #102.

**Comment by thw26** [comment]
Attempting to test a function to test the dialog version and hitting an error when running `./main.py -I`. Hold on review.

**Comment by thw26** [comment]
Fixed the issue. It appears Logos introduced a BOM in their XML data.

**Comment by thw26** [comment]
Fixed `-I`. Now I'm getting errors on my dialog version checker. Progress!

**Comment by thw26** [comment]
And the test is now fixed. Should be ready for review. Again, the conflicts are due to a cherry-pick followed by a rebase.

**Comment by thw26** [comment]
Dropped the XML BOM commit and fixed the cherry-pick/rebase issue.

**Comment by thw26** [comment]
And I have reworked the `dialog` package detection so that it should fallback to curses if the dialog package is outdated. This may break the installer but since I cannot test it at present, I am assuming it will work. The one place I expect it to break is when choosing the install path.

File: installer.py [hunk 1]
```diff
@@ -19,27 +17,11 @@ def ensure_product_choice(app=None):
     logging.debug('- config.FLPRODUCTi')
     logging.debug('- config.VERBUM_PATH')
 
-    if not config.FLPRODUCT:
```
Comment (n8marti) [inline installer.py:22]: The reason for having the `if not config.VARIABLE` statements at each install step was to be able to skip the step if the variable was already set. Without them each step gets processed in full, even if the user has already set the variable, either in the config file or with env variables.

Comment (thw26) [reply installer.py:22]: I believe I commented those out from an earlier revision, before I had set up proper threading in the TUI. Those should be easy to restore.

**Comment by thw26** [comment]
Most recent push simplifies tui_app.set_tui_menu_options().

**Comment by thw26** [comment]
Most recent push should fix process sharing between main.py and wine.py.

**Comment by n8marti** [comment]
Success! (mostly) I assume the following error is due to ICU files:
image

I'm going to test again with 29.x.

**Comment by n8marti** [comment]
Well, I got the same error, except I got it because the GUI still installs 34.x instead of the 29.x that I selected. That's a separate issue due to the config file's release version for some reason taking precedence over the once chosen in the GUI. See [here](https://github.com/FaithLife-Community/LogosLinuxInstaller/pull/118#issuecomment-2202365819). After removing the config file and trying again it works.

In short, I think we're ready to merge this in.
